### ISSUE TITLE
redundant_null_checks_lint_rule

### ISSUE THREAD

**Issue by jacob314** [dart-lang/linter#2783]
**Describe the rule you'd like to see implemented**

Warn about  duplicate null checks that can be eliminated by assigning expressions to local variables.

Ideally provide a quick fix that uses the existing *extract local refactoring* combined with stripping out null checks that aren't needed.

**Examples**

# 
```dart
if (stmt.sourceSpan != null) {
  sourceComment = '${stmt.sourceSpan!.sourceUrl}';
}
```
Warns that the null checks part of `stmt.sourceSpan != null` and `stmt.sourceSpan!` are redundant and offers the quick fix to 

```
var sourceSpan = stmt.sourceSpan;
if (sourceSpan != null) {
  sourceComment = '${sourceSpan.sourceUrl}';
}
```

Also applies for cases with more than 2 redundant null checks such as 
```dart
if (stmt.sourceSpan != null) {
      sourceComment = ' /* '
          '${stmt.sourceSpan!.sourceUrl}'
          ':'
          '${stmt.sourceSpan!.start.offset}'
          ':'
          '${stmt.sourceSpan!.start.offset}'
          ' */';
}
```
Should warn about 4 redundant null checks and offer a quick fix to:
```dart
var sourceSpan = stmt.sourceSpan;
if (sourceSpan != null) {
      sourceComment = ' /* '
          '${sourceSpan.sourceUrl}'
          ':'
          '${sourceSpan.start.offset}'
          ':'
          '${sourceSpan.start.offset}'
          ' */';
}
```

**Additional context**
Users don't understand that it isn't safe for Dart to trust that the value of `foo.bar` does not change between multiple times it is accessed. On the other hand, users write code assuming the value will not change between multiple calls.

**Comment by jakemac53** [comment]
Note that this lint should also probably check for any assignments to that property, and transform them like so:

```dart
// Before
stmt.sourceSpan = <something>;
// After
sourceSpan = stmt.sourceSpan = <something>;
```

Or maybe just refuse to do the refactor

**Comment by bwilkerson** [comment]
If there are assignments to the property between the condition and the access then it should not produce a lint.

We should run the lint over a large corpus of code in order to estimate the number of false positives.

**Comment by stuartmorgan** [comment]
> If there are assignments to the property between the condition and the access then it should not produce a lint.

There are common patterns, such as a lazily constructed member variable, where is would be far better to do the assignment refactor than leave it alone. This was a common pain point during NNBD migrations, and Peele didn't always get it right when adding a local manually.

**Comment by bwilkerson** [comment]
Could you provide an example of the pattern you're thinking of?

**Comment by stuartmorgan** [comment]
I'd have to try to dig through old emails to find a real example, but roughly things that used to be (pre-NNBD):

```
Foo _foo;
Foo get foo {
  if (_foo == null) {
     // Do some stuff here
     _foo = resultOfAbove;
  }
  return _foo;
}
```

The native null safe conversion if the goal is non-nullability of the getter (assuming a case where `late` isn't an option, such as when there are cases that re-null `_foo`) is:
```
Foo? _foo;
Foo get foo {
  if (_foo == null) {
     // Do some stuff here
     _foo = resultOfAbove;
  }
  return _foo!;
}
```

That's dangerous because the compiler isn't enforcing that the thing beyond assigned to `_foo` is not null. Introducing a local addresses that, but I saw people failing to correctly update `_foo`, so that the intended cache didn't actually work. An automated refactor would avoid the human error.

**Comment by bwilkerson** [comment]
Thanks, that helps. Yes, lazy initialization is more difficult under null safety. It's possible, though, that we might be able to do something even better if we split the lazy initialization case out from the originally proposed lint. I still think that for every case other than lazy initialization we don't want to produce a lint if the field is assigned a value because it's too easy to introduce bugs in those cases. If you know of other cases I'd be interested to hear about them.

But for the lazy initialization case, what if we were to introduce an annotation (I'd call it `@lazy`, but that's too similar to the keyword) that could be used to enforce the pattern that we settle on for how best to lazily initialize a field under null safety. It could potentially enforce other conventions, such as not reading the value of the field except through the corresponding getter. It could still allow assignments for the case where the value needs to be flushed (though perhaps it only allows `null` to be assigned when it's the literal value and not when it's coming from a nullable variable), or maybe control that checking with a parameter to the annotation. (Maybe we're really talking about a simple cached value and the annotation should be `@cached`.)

In case it's interesting, in the analyzer code base we had a few lazily initialized values like this and we decided to make the field non-nullable in order to simplify the problem. We did that by choosing a sentinel value other than `null` and initialized the field to that value. The getter then checks for the sentinel value to determine whether the field needs to be initialized. It still has the same disadvantages you always have with using an invalid value, but it gets around the need for explicit null checks.

**Comment by jacob314** [comment]
We would need to align this with the style guide which currently suggests ["Sometimes it’s best to simply use ! on the field."](https://dart.dev/guides/language/effective-dart/usage#consider-assigning-a-nullable-field-to-a-local-variable-to-enable-type-promotion)

**Comment by davidmorgan** [comment]
I'm not sure this works as a lint.

The refactoring is behavior changing; there's no way for the lint to know if it's correct. People tend to just follow what a lint says, so lints that sometimes suggest the wrong thing are a problem.

The line for what's desirable for style is also not super clear, which is another argument against a lint.

The quick fix would be great to have though.
